scylla.go

```diff
@@ -331,3 +394,259 @@ func (p *scyllaConnPicker) randomConn() *Conn {
+
+	if time.Now().Before(disabledUntil) {
+		// Temporarily disabled, fallback to the old method
+		if gocqlDebug {
+			Logger.Printf("shard aware port disabled until %v, falling back to the regular one", disabledUntil)
+		}
+		return nil
+	}
+
+	var targetAddress string
+	if cfg.tlsConfig != nil {
+		targetAddress = p.shardAwareAddressSSL
+	} else {
+		targetAddress = p.shardAwareAddress
+	}
+
+	if targetAddress == "" {
+		// Shard-aware port not supported by the host, fallback to the old method
+		if gocqlDebug {
+			Logger.Printf("no shard-aware port info for SSL=%b, falling back to regular port", cfg.tlsConfig != nil)
+		}
+		return nil
+	}
+
+	shard, ok := p.getShardToConnect()
+	if !ok {
+		// For each shard, a connection is being established or is already open
+		if gocqlDebug {
+			Logger.Printf("did not get a shard to connect, falling back to regular port")
+		}
+		return nil
+	}
+
+	portPicker := newScyllaPortPicker(shard, p.nrShards)
+	sourcePort, hasPort := portPicker.nextPort()
+	if !hasPort {
+		// There are no ports in range?
+		return nil
+	}
+
+	return &scyllaShardAwarePortOptions{
+		targetAddress: targetAddress,
+		sourcePort:    sourcePort,
+		shard:         shard,
+		portPicker:    portPicker,
+	}
+}
+
+type scyllaConnParams struct {
+	*scyllaShardAwarePortOptions
+
+	connPicker *scyllaConnPicker
+}
+
+// must be called under hostConnPool's lock
+func (scp *scyllaConnParams) verifyConnection(conn *Conn) {
+	if !scp.usesShardAwarePort() {
+		return
+	}
+
+	if conn.scyllaSupported.shard != scp.getShard() {
+		// Something's wrong, we got a connection to the wrong shard.
+		// It may be caused by network configuration issues (e.g. NAT
+		// modifies the source port).
+		// Temporarily disable connecting to the shard aware port. This will allow
+		// us to make progress in establishing connection per shard, and also
+		// make it possible to use the new method if network configuration is fixed
+		// in the meantime.
+		scp.connPicker.disableShardAwareConnectionsUntil = time.Now().Add(1 * time.Minute)
+
+		if gocqlDebug {
+			Logger.Printf("disabling connections to shard aware port for host %s, %s",
+				scp.connPicker.shardAwareAddress, scp.connPicker.shardAwareAddressSSL)
+		}
+	}
+}
+
+func (scp *scyllaConnParams) usesShardAwarePort() bool {
+	return scp != nil && scp.scyllaShardAwarePortOptions != nil
+}
+
+func (scp *scyllaConnParams) updateToNextPort() bool {
+	if !scp.usesShardAwarePort() {
+		return false
+	}
+	return scp.scyllaShardAwarePortOptions.updateToNextPort()
+}
+
+func (scp *scyllaConnParams) getShard() int {
+	if !scp.usesShardAwarePort() {
+		return -1
+	}
+	return scp.scyllaShardAwarePortOptions.shard
+}
+
+type scyllaShardAwarePortOptions struct {
+	targetAddress string
+	sourcePort    uint16
+	shard         int
+
+	portPicker *scyllaPortPicker
+}
+
+func (ssapo *scyllaShardAwarePortOptions) updateToNextPort() bool {
+	var hadPort bool
+	ssapo.sourcePort, hadPort = ssapo.portPicker.nextPort()
```

Comment:
If sth is called options or params should be immutable. This is super ugly. The portPicker should not be part of scyllaShardAwarePortOptions. This is some kind of iterator.